### PR TITLE
Use a more typesafe approach for managing indexed data

### PR DESCRIPTION
**Pull request by tifecool**
Updated fix for issue: #2403

Please review if this is a feasible solution. @ordian 

polkadot address: 1s51dYXziQwTnSnModErKfra3TeEsVipqzY846CLCTEXtHJ

### CODE REVIEW

File: node/subsystem-util/src/rolling_session_window.rs [hunk 1]
```diff
@@ -392,17 +392,17 @@ mod tests {
...
 	pub const TEST_WINDOW_SIZE: SessionWindowSize = new_session_window_size!(6);
 
 	fn dummy_session_info(index: SessionIndex) -> SessionInfo {
 		SessionInfo {
-			validators: Vec::new(),
+			validators: Validators::from(Vec::new()),
 			discovery_keys: Vec::new(),
 			assignment_keys: Vec::new(),
-			validator_groups: Vec::new(),
+			validator_groups: GroupValidators::from(Vec::new()),
```
Comment (sandreim) [inline node/subsystem-util/src/rolling_session_window.rs:405]: I think impl Default would be better.

File: node/subsystem-util/src/runtime/mod.rs [hunk 2]
```diff
@@ -207,10 +207,10 @@ impl RuntimeInfo {
 	///
 	/// Returns: `None` if not a parachain validator.
 	async fn get_validator_info(&self, session_info: &SessionInfo) -> Result<ValidatorInfo> {
-		if let Some(our_index) = self.get_our_index(&session_info.validators).await {
+		if let Some(our_index) = self.get_our_index(&session_info.validators.to_vec()).await {
 			// Get our group index:
 			let our_group =
-				session_info.validator_groups.iter().enumerate().find_map(|(i, g)| {
+				session_info.validator_groups.to_vec().iter().enumerate().find_map(|(i, g)| {
```
Comment (sandreim) [inline node/subsystem-util/src/runtime/mod.rs:213]: This would create an additional clone.

File: node/subsystem-util/src/runtime/mod.rs [hunk 3]
```diff
@@ -207,10 +207,10 @@ impl RuntimeInfo {
 	///
 	/// Returns: `None` if not a parachain validator.
 	async fn get_validator_info(&self, session_info: &SessionInfo) -> Result<ValidatorInfo> {
-		if let Some(our_index) = self.get_our_index(&session_info.validators).await {
+		if let Some(our_index) = self.get_our_index(&session_info.validators.to_vec()).await {
```
Comment (sandreim) [inline node/subsystem-util/src/runtime/mod.rs:210]: We could update `get_our_index` to work with the new type `Validators` which would also require to implement an iterator interface.

File: primitives/src/v2/mod.rs [hunk 4]
```diff
@@ -1569,6 +1569,78 @@ impl CompactStatement {
...
+/// Validators struct indexed by ValidatorIndex.
+#[derive(Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+#[cfg_attr(feature = "std", derive(PartialEq, MallocSizeOf))]
+pub struct Validators(Vec<ValidatorId>);
+
+impl Index<ValidatorIndex> for Validators {
+	type Output = ValidatorId;
+
+	fn index(&self, index: ValidatorIndex) -> &Self::Output {
+		&self.0.get(index.0 as usize).unwrap()
```
Comment (sandreim) [inline primitives/src/v2/mod.rs:1581]: This looks dangerous. Should `Output` be option ?

File: primitives/src/v2/mod.rs [hunk 5]
```diff
@@ -1569,6 +1569,78 @@ impl CompactStatement {
...
+	}
+}
+
+impl GroupValidators {
+	/// Returns a reference to an element indexed using GroupIndex.
+	pub fn get(&self, index: GroupIndex) -> Option<&Vec<ValidatorIndex>> {
+		self.0.get(index.0 as usize)
+	}
+
+	///Returns number of elements in vector.
```
Comment (sandreim) [inline primitives/src/v2/mod.rs:1633]:
```suggestion
	/// Returns number of elements in vector.
```

File: primitives/src/v2/mod.rs [hunk 6]
```diff
@@ -1569,6 +1569,78 @@ impl CompactStatement {
...
+	pub fn len(&self) -> usize {
+		self.0.len()
+	}
+
+	///Returns contained vector
```
Comment (sandreim) [inline primitives/src/v2/mod.rs:1638]:
```suggestion
	/// Returns contained vector
```

File: primitives/src/v2/mod.rs [hunk 7]
```diff
@@ -1569,6 +1569,78 @@ impl CompactStatement {
...
+	pub fn get(&self, index: ValidatorIndex) -> Option<&ValidatorId> {
+		self.0.get(index.0 as usize)
+	}
+
+	///Returns number of elements in vector.
+	pub fn len(&self) -> usize {
+		self.0.len()
+	}
+
+	///Returns contained vector
```
Comment (sandreim) [inline primitives/src/v2/mod.rs:1602]:
```suggestion
	/// Returns contained vector
```

File: primitives/src/v2/mod.rs (hunk 4 above)
Comment (tifecool) [reply primitives/src/v2/mod.rs:1581]: I implemented it just in case. 
Seen a few places where the original were indexed using e.g. `validators[0]` syntax

**Comment by tifecool** [comment]
Will look to a cleaner solution using generics.
Too much repeated code.

File: primitives/src/v2/mod.rs (hunk 4 above)
Comment (tifecool) [reply primitives/src/v2/mod.rs:1586]: >  Should Output be option ?

Tried this approach, couldn't quite hack it

**Comment by tifecool** [comment]
@eskimor is this a workable approach?

File: primitives/src/v2/mod.rs [hunk 8]
```diff
@@ -1569,6 +1574,140 @@ impl CompactStatement {
...
+	/// Returns true if contained Vector is empty
+	pub fn is_empty(&self) -> bool {
+		self.0.is_empty()
+	}
+}
+
+/// GroupValidators struct indexed by GroupIndex.
+#[derive(Clone, Encode, Decode, RuntimeDebug, TypeInfo, Default)]
+#[cfg_attr(feature = "std", derive(PartialEq, MallocSizeOf))]
+pub struct GroupValidators(Vec<Vec<ValidatorIndex>>);
```
Comment (ordian) [inline primitives/src/v2/mod.rs:1647]: I'd suggest to name it `ValidatorGroups`.

File: primitives/src/v2/mod.rs (hunk 4 above)
Comment (ordian) [reply primitives/src/v2/mod.rs:1586]: Agree this looks as dangerous as using `validators[index.0 as usize]` which panics for a vector on out-of-bounds access.

I'd prefer to remove this implementation completely and replace the usage of indexing with `get(index).unwrap()` in tests and `get(index).expect("<write an actual proof here; qed>")` elsewhere.

File: primitives/src/v2/mod.rs [hunk 9]
```diff
@@ -1569,6 +1574,140 @@ impl CompactStatement {
...
+}
+
+impl FromIterator<ValidatorId> for Validators {
+	fn from_iter<T: IntoIterator<Item = ValidatorId>>(iter: T) -> Self {
+		let mut validators = Vec::new();
+
+		for i in iter {
+			validators.push(i);
+		}
+		Validators(validators)
```
Comment (ordian) [inline primitives/src/v2/mod.rs:1603]:
```suggestion
		Self(Vec::from_iter(iter))
```

File: primitives/src/v2/mod.rs [hunk 10]
```diff
@@ -1569,6 +1574,140 @@ impl CompactStatement {
...
+}
+
+impl FromIterator<Vec<ValidatorIndex>> for GroupValidators {
+	fn from_iter<T: IntoIterator<Item = Vec<ValidatorIndex>>>(iter: T) -> Self {
+		let mut validator_groups = Vec::new();
+
+		for i in iter {
+			validator_groups.push(i);
+		}
+		GroupValidators(validator_groups)
```
Comment (ordian) [inline primitives/src/v2/mod.rs:1670]:
```suggestion
		Self(Vec::from_iter(iter))
```

File: primitives/src/v2/mod.rs [hunk 11]
```diff
@@ -1569,6 +1574,140 @@ impl CompactStatement {
...
+	}
+}
+
+impl GroupValidators {
+	/// Returns a reference to an element indexed using GroupIndex.
```
Comment (ordian) [inline primitives/src/v2/mod.rs:1675]: Please wrap types in comments in `` to satisfy spellchecker, e.g. `GroupIndex`.

File: primitives/src/v2/mod.rs [hunk 12]
```diff
@@ -1569,6 +1574,140 @@ impl CompactStatement {
...
+	pub fn get(&self, index: GroupIndex) -> Option<&Vec<ValidatorIndex>> {
+		self.0.get(index.0 as usize)
+	}
+
+	/// Returns number of elements in vector.
```
Comment (ordian) [inline primitives/src/v2/mod.rs:1680]:
```suggestion
	/// Returns the number of validator groups.
```

File: primitives/src/v2/mod.rs [hunk 13]
```diff
@@ -1569,6 +1574,140 @@ impl CompactStatement {
...
+	pub fn iter_mut(&mut self) -> IterMut<'_, Vec<ValidatorIndex>> {
+		self.0.iter_mut()
+	}
+
+	/// Returns an Iterator with ownership.
+	pub fn into_iter(self) -> IntoIter<Vec<ValidatorIndex>> {
+		self.0.into_iter()
+	}
+
+	/// Returns true if contained Vector is empty
```
Comment (ordian) [inline primitives/src/v2/mod.rs:1705]:
```suggestion
	/// Returns true if the underlying container is empty.
```

File: primitives/src/v2/mod.rs [hunk 14]
```diff
@@ -1569,6 +1574,140 @@ impl CompactStatement {
...
+	pub fn iter(&self) -> Iter<'_, Vec<ValidatorIndex>> {
+		self.0.iter()
+	}
+
+	/// Returns a mutable Iterator.
+	pub fn iter_mut(&mut self) -> IterMut<'_, Vec<ValidatorIndex>> {
+		self.0.iter_mut()
+	}
+
+	/// Returns an Iterator with ownership.
```
Comment (ordian) [inline primitives/src/v2/mod.rs:1700]:
```suggestion
	/// Creates a consuming iterator.
```

File: primitives/src/v2/mod.rs [hunk 15]
```diff
@@ -1569,6 +1574,140 @@ impl CompactStatement {
...
+	pub fn to_vec(&self) -> Vec<Vec<ValidatorIndex>> {
+		self.0.clone()
+	}
+
+	/// Returns Iterator of contained vector.
+	pub fn iter(&self) -> Iter<'_, Vec<ValidatorIndex>> {
+		self.0.iter()
+	}
+
+	/// Returns a mutable Iterator.
```
Comment (ordian) [inline primitives/src/v2/mod.rs:1695]:
```suggestion
	/// Returns an iterator that allows modifying each group.
```

File: primitives/src/v2/mod.rs [hunk 16]
```diff
@@ -1569,6 +1574,140 @@ impl CompactStatement {
...
+	pub fn len(&self) -> usize {
+		self.0.len()
+	}
+
+	/// Returns contained vector.
+	pub fn to_vec(&self) -> Vec<Vec<ValidatorIndex>> {
+		self.0.clone()
+	}
+
+	/// Returns Iterator of contained vector.
```
Comment (ordian) [inline primitives/src/v2/mod.rs:1690]:
```suggestion
	/// Returns an iterator over the groups.
```

File: primitives/src/v2/mod.rs [hunk 17]
```diff
@@ -1569,6 +1573,114 @@ impl CompactStatement {
...
+	}
+}
+
+impl FromIterator<ValidatorId> for Validators {
+	fn from_iter<T: IntoIterator<Item = ValidatorId>>(iter: T) -> Self {
+		Self(Vec::from_iter(iter))
+	}
+}
+
+impl Validators {
```
Comment (eskimor) [inline primitives/src/v2/mod.rs:1593]: We also have other keys indexed by `ValidatorIndex`, something more generic might make sense here with `Validators` being a type alias for something like `IndexedVec<ValidatorIndex, ValidatorId>`.

Other than that looks really good at first sight.

Comment (tifecool) [reply primitives/src/v2/mod.rs:1593]: Actually considering the generic version. Will look to implementing it.

File: node/network/statement-distribution/src/lib.rs [hunk 18]
```diff
@@ -2072,7 +2072,10 @@ impl<R: rand::Rng> StatementDistributionSubsystem<R> {
 					// directly:
 					let group_peers = {
 						if let Some(our_group) = validator_info.our_group {
-							let our_group = &session_info.validator_groups[our_group.0 as usize];
+							let our_group = &session_info
+								.validator_groups
+								.get(our_group)
+								.expect("<write an actual proof here; qed>");
```
Comment (ordian) [inline node/network/statement-distribution/src/lib.rs:2078]:
```suggestion
								.expect("`our_group` is derived from `validator_groups`; qed");
```

File: node/core/dispute-coordinator/src/import.rs [hunk 19]
```diff
@@ -488,7 +488,7 @@ impl ImportResult {
 		for (index, sig) in approval_votes.into_iter() {
 			debug_assert!(
 				{
-					let pub_key = &env.session_info().validators[index.0 as usize];
+					let pub_key = &env.session_info().validators.get(index).expect("<write an actual proof here; qed>");
```
Comment (ordian) [inline node/core/dispute-coordinator/src/import.rs:491]: :sweat_smile: I didn't mean `write an actual proof here` quite literally. If you grep for [`qed`](https://cs.github.com/paritytech/polkadot?q=qed), you'll see what kind of proof I meant. In this case it would be something like:
```suggestion
					let pub_key = &env.session_info().validators.get(index).expect("indices are validated by approval-voting subsystem; qed");
```

File: node/core/dispute-coordinator/src/initialized.rs [hunk 20]
```diff
@@ -1097,7 +1097,7 @@ impl Initialized {
 				valid,
 				candidate_hash,
 				session,
-				env.validators()[index.0 as usize].clone(),
+				env.validators().get(*index).expect("<write an actual proof here; qed>").clone(),
```
Comment (ordian) [inline node/core/dispute-coordinator/src/initialized.rs:1100]:
```suggestion
				env.validators().get(*index).expect("`controlled_indices` are derived from `validators`; qed").clone(),
```

File: primitives/src/v2/mod.rs [hunk 21]
```diff
@@ -1569,6 +1594,72 @@ impl CompactStatement {
...
+
+	/// Returns Iterator of contained vector.
+	pub fn iter(&self) -> Iter<'_, V> {
+		self.0.iter()
+	}
+
+	/// Returns a mutable Iterator.
+	pub fn iter_mut(&mut self) -> IterMut<'_, V> {
+		self.0.iter_mut()
+	}
```
Comment (ordian) [inline primitives/src/v2/mod.rs:1650]: Is it used anywhere? If not, I'd suggest deleting.

Comment (tifecool) [reply primitives/src/v2/mod.rs:1650]: Strange, thought it was used somewhere. Will remove

Comment (tifecool) [reply primitives/src/v2/mod.rs:1650]: It's used `node/network/availability-distribution/src/requester/session_cache.rs` line 195

File: primitives/src/v2/mod.rs [hunk 22]
```diff
@@ -1569,6 +1594,72 @@ impl CompactStatement {
 	}
 }
 
+/// `IndexedVec` struct indexed by type specific indexes.
```
Comment (ordian) [inline primitives/src/v2/mod.rs:1597]: Both `indexes` and `indices` are correct, but we prefer the latter in this codebase
```suggestion
/// `IndexedVec` struct indexed by type specific indices.
```

File: primitives/src/v2/mod.rs [hunk 23]
```diff
@@ -1569,6 +1594,72 @@ impl CompactStatement {
...
+	pub fn to_vec(&self) -> Vec<V> {
+		self.0.clone()
+	}
+
+	/// Returns an iterator over the underlying vector.
+	pub fn iter(&self) -> Iter<'_, V> {
+		self.0.iter()
+	}
+
+	/// Returns a mutable Iterator.
```
Comment (ordian) [inline primitives/src/v2/mod.rs:1647]:
```suggestion
	/// Returns a mutable iterator over the underlying vector.
```

File: primitives/src/v2/mod.rs [hunk 24]
```diff
@@ -790,6 +809,12 @@ impl From<u32> for GroupIndex {
 	}
 }
 
+impl From<GroupIndex> for u32 {
+	fn from(i: GroupIndex) -> Self {
+		i.0
+	}
+}
+
```
Comment (ordian) [inline primitives/src/v2/mod.rs:817]: What is the motivation to add these? The reason we have typed indices is to ensure type-safety, that is we don't accidentally index a vector of groups with a validator index. With these conversions to u32, it's easy to make a mistake with `.into()`.

File: primitives/src/v2/mod.rs [hunk 25]
```diff
@@ -124,7 +129,9 @@ impl MallocSizeOf for ValidatorId {
 }
 
 /// Index of the validator is used as a lightweight replacement of the `ValidatorId` when appropriate.
-#[derive(Eq, Ord, PartialEq, PartialOrd, Copy, Clone, Encode, Decode, TypeInfo, RuntimeDebug)]
+#[derive(
+	Eq, Ord, PartialEq, PartialOrd, Copy, Clone, Encode, Decode, TypeInfo, RuntimeDebug, Default,
```
Comment (ordian) [inline primitives/src/v2/mod.rs:133]: Why did you add `Default`? We're trying to be very strict with types, especially in the runtime. It's not clear to me what `ValidatorIndex(0)` being `default` actually means.

File: primitives/src/v2/mod.rs [hunk 26]
```diff
@@ -1569,6 +1594,72 @@ impl CompactStatement {
 	}
 }
 
+/// `IndexedVec` struct indexed by type specific indexes.
+#[derive(Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+#[cfg_attr(feature = "std", derive(PartialEq, MallocSizeOf))]
+pub struct IndexedVec<K, V>(Vec<V>, PhantomData<fn(K) -> K>);
```
Comment (ordian) [inline primitives/src/v2/mod.rs:1600]: Could someone confirm this has the same SCALE (d)encoding as a regular `Vec`?

**Review by ordian — COMMENTED** [review]
Thank you! Just some final comments.

File: primitives/src/v2/mod.rs (hunk 25 above)
Comment (tifecool) [reply primitives/src/v2/mod.rs:133]: Added it when I was trying to derive Default for `IndexedVec`. Unnecessary now I implemented a default.

File: primitives/src/v2/mod.rs (hunk 24 above)
Comment (tifecool) [reply primitives/src/v2/mod.rs:817]: Using another approach which should work better.

File: primitives/src/v2/mod.rs (hunk 26 above)
Comment (bkchr) [reply primitives/src/v2/mod.rs:1604]: Yes it has the same. `PhantomData` encodes into "nothing" and decodes from "nothing"

File: primitives/src/v2/mod.rs [hunk 27]
```diff
@@ -1569,6 +1598,72 @@ impl CompactStatement {
 	}
 }
 
+/// `IndexedVec` struct indexed by type specific indices.
+#[derive(Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+#[cfg_attr(feature = "std", derive(PartialEq, MallocSizeOf))]
+pub struct IndexedVec<K, V>(Vec<V>, PhantomData<fn(K) -> K>);
```
Comment (bkchr) [inline primitives/src/v2/mod.rs:1604]:
```suggestion
pub struct IndexedVec<K, V>(Vec<V>, PhantomData<K>);
```

File: primitives/src/v2/mod.rs [hunk 28]
```diff
@@ -1569,6 +1598,72 @@ impl CompactStatement {
...
+		Self(Vec::from_iter(iter), PhantomData)
+	}
+}
+
+impl<K, V> IndexedVec<K, V>
+where
+	V: Clone,
+{
+	/// Returns a reference to an element indexed using `K`.
+	pub fn get(&self, index: K) -> Option<&V>
```
Comment (bkchr) [inline primitives/src/v2/mod.rs:1629]: Why not take `K` by reference?

File: primitives/src/v2/mod.rs [hunk 29]
```diff
@@ -1569,6 +1598,72 @@ impl CompactStatement {
...
+
+	/// Creates a consuming iterator.
+	pub fn into_iter(self) -> IntoIter<V> {
+		self.0.into_iter()
+	}
+
+	/// Returns true if the underlying container is empty.
+	pub fn is_empty(&self) -> bool {
+		self.0.is_empty()
+	}
```
Comment (bkchr) [inline primitives/src/v2/mod.rs:1664]: You can just implement `Deref` to get all these functions for "free"

Only `get` function is required and the compiler will then figure out the correct version to call.

File: primitives/src/v2/mod.rs [hunk 30]
```diff
@@ -123,6 +128,12 @@ impl MallocSizeOf for ValidatorId {
 	}
 }
 
+/// Trait required for type specific indices e.g. `ValidatorIndex` and `GroupIndex`
+pub trait TypeIndex {
+	///Converts `Self` to `usize`.
+	fn into_usize(self) -> usize;
```
Comment (bkchr) [inline primitives/src/v2/mod.rs:134]:
```suggestion
	/// Returns the index associated to this value.
	fn type_index(&self) -> usize;
```

Or maybe something different, but `into_usize` is really opaque.

File: primitives/src/v2/mod.rs (hunk 27 above)
Comment (ordian) [reply primitives/src/v2/mod.rs:1604]: https://docs.rs/typed-index-collections/3.1.0/src/typed_index_collections/vec.rs.html#89-109

File: primitives/src/v2/mod.rs (hunk 29 above)
Comment (ordian) [reply primitives/src/v2/mod.rs:1664]: We want to ban indexing operations those and require using `.get(index)`.

File: primitives/src/v2/mod.rs (hunk 30 above)
Comment (ordian) [reply primitives/src/v2/mod.rs:134]: This trait should be private if possible.

Comment (tifecool) [reply primitives/src/v2/mod.rs:134]: Had to be public, had leaking error.

Comment (bkchr) [reply primitives/src/v2/mod.rs:134]: It appears in the public interface, so it needs to be public.

File: primitives/src/v2/mod.rs (hunk 28 above)
Comment (tifecool) [reply primitives/src/v2/mod.rs:1629]: No reason, just stuck with it. Can refactor if necessary.

File: primitives/src/v2/mod.rs (hunk 27 above)
Comment (bkchr) [reply primitives/src/v2/mod.rs:1604]: Okay, you send me down a rabbit hole :P Do you know why `K` as invariant is not appearing in the auto trait derives? Just some Rustc magic of ignoring?

**Review by ordian — APPROVED** [review]
We can always bikeshed further, but it looks good to me as is.

**Comment by ordian** [comment]
I'd suggest a medium tip given the work spent in #5503 and here.

cc @kianenigma @shawntabrizi

File: primitives/src/v2/mod.rs (hunk 27 above)
Comment (ordian) [reply primitives/src/v2/mod.rs:1604]: Sorry :) 
I think auto trait derives do not suffer some regular dumb `derive(Trait)` limitations that every type mentioned must `Trait`.
In this case `fn(K) -> K` impls `Send` even if `K` doesn't: https://doc.rust-lang.org/core/primitive.fn.html#synthetic-implementations. I don't know where is this is defined though. But doesn't seem to be in
https://github.com/rust-lang/rust/blob/master/compiler/rustc_trait_selection/src/traits/auto_trait.rs

Comment (ordian) [reply primitives/src/v2/mod.rs:1604]: Hmm. So does `fn(K)`. I'm not quire sure what is the point of `fn(K) -> K` then.
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9510d8875227f41600b6c871f7394f91

Comment (bkchr) [reply primitives/src/v2/mod.rs:1604]: Haha. No problem. Seems like something interested where you can waste a lot of time on 🤣

Comment (tifecool) [reply primitives/src/v2/mod.rs:1604]: Tbh, I just used the code directly as I saw it... Considered `PhantomData<K>` but didn't want to have any weird issues while testing the generics.

**Comment by tifecool** [comment]
I did see some other types that could benefit from the new generic Struct e.g. discovery_keys.

Let me know if others should implement the IndexedVec type.

**Review by bkchr — APPROVED** [review]

**Comment by bkchr** [comment]
bot merge

**Comment by eskimor** [comment]
Awesome, thank you!
